### DISCUSSION TITLE
Has Quarto 1.4 become slower than 1.3 with rendering PDF books with flextables?

### DISCUSSION THREAD

**Discussion by liyongg** [quarto-dev/quarto-cli#9695] [Q&A]
### Description

Our team at work has transitioned from Word to a Quarto book project using R to ultimately publish our annual report as PDF. This was done around December 2023 and January 2024, in which we used Quarto 1.3. The book project consists of 8 `.qmd` files and has a pre- and post-render script to mutate some things in the output `.tex` file, which then is rendered with `tinytex::xelatex()` in the post-render script. The report has a lot of `ggplot` and `flextable` objects.

Now I am preparing for next year/round and I am checking for compatibility issues with rendering the book project using Quarto 1.4, as I am not sure if Quarto 1.5 will already be available by then. I am using [qvm](https://github.com/dpastoor/qvm) to install multiple versions of Quarto on my Windows client. There seem to be differences in render time for both versions. Excuse me for not having the time yet to bring a reprex. I hope that some pointers could help me resolve the problem, and will try to make time for a reprex if absolutely necessary. Last time I tried, the flextables I created did not cause these huge differences. For what it's worth, I have rendered the Quarto project in two independent environments: on a VM provided by work, but also on a separate device (used by the team for work) and experienced/observed similar differences in render time between Quarto versions.

## OS and `quarto check` output

`Windows 10 x64`

```
$ quarto1.3 check

[>] Checking versions of quarto binary dependencies...
      Pandoc version 3.1.1: OK
      Dart Sass version 1.55.0: OK
[>] Checking versions of quarto dependencies......OK
[>] Checking Quarto installation......OK
      Version: 1.3.450
      Path: C:\Users\x\AppData\Local\qvm\versions\v1.3.450\bin
      CodePage: 1252

[>] Checking basic markdown render....OK

[>] Checking Python 3 installation....OK
      Version: 3.9.5
      Path: C:/Users/x/AppData/Local/Programs/Python/Python39/python.exe
      Jupyter: (None)

      Jupyter is not available in this Python installation.
      Install with py -m pip install jupyter

(/) Checking R installation...........✔ .Rprofile: Rdata file ingelezen:  data/RR2024_20240117_131326.Rdata
[>] Checking R installation...........(None)

      Unable to locate an installed version of R.
      Install R from https://cloud.r-project.org/
```

```
$ quarto check
Quarto 1.4.554
[>] Checking versions of quarto binary dependencies...
      Pandoc version 3.1.11: OK
      Dart Sass version 1.69.5: OK
      Deno version 1.37.2: OK
[>] Checking versions of quarto dependencies......OK
[>] Checking Quarto installation......OK
      Version: 1.4.554
      Path: C:\Users\x\AppData\Local\qvm\versions\v1.4.554\bin
      CodePage: 1252

[>] Checking tools....................OK
      TinyTeX: v2024.04
      Chromium: (not installed)

[>] Checking LaTeX....................OK
      Using: TinyTex
      Path: C:\Users\x\AppData\Roaming\TinyTeX\bin\windows\
      Version: 2024

[>] Checking basic markdown render....OK

[>] Checking Python 3 installation....OK
      Version: 3.9.5
      Path: C:/Users/x/AppData/Local/Programs/Python/Python39/python.exe
      Jupyter: (None)

      Jupyter is not available in this Python installation.
      Install with py -m pip install jupyter

(/) Checking R installation...........✔ .Rprofile: Rdata file ingelezen:  data/RR2024_20240117_131326.Rdata
[>] Checking R installation...........OK
      Version: 4.2.3
      Path: C:/PROGRA~1/R/R-42~1.3
      LibPaths:
        - C:/Tools/proj/renv/library/R-4.2/x86_64-w64-mingw32
        - C:/Users/x/AppData/Local/R/cache/R/renv/sandbox/R-4.2/x86_64-w64-mingw32/19d6eec0
      knitr: 1.45
      rmarkdown: 2.25

[>] Checking Knitr engine render......OK
```

## Only text and inline R code
I have tried quite some debugging and noticed that v1.3 has approximately the same render times as v1.4 when excluding `.qmd` files that have use quite a lot of `flextable` and `ggplot` objects. But it does contain some R code chunks and also uses the inline variable syntax, `r variable_defined_in_chunk`

```
$ time quarto1.3 render
...
real    1m22,317s
user    0m0,031s
sys     0m0,046s
```

```
$ time quarto render
real    1m22,825s
user    0m0,015s
sys     0m0,046s
```

This is reasonable!

## Including one additional `.qmd` file with many `flextable` and `ggplot` objects
```
$ time quarto1.3 render
...
real    2m20,554s
user    0m0,015s
sys     0m0,061s
```

```
$ time quarto render
...
real    15m14,170s
user    0m0,015s
sys     0m0,062s
```

Not sure if this is any helpful, but I noticed it takes a lot longer before the post-render script starts. In Quarto 1.4 it seems to hang at:

```
pandoc
...
  date: 12 januari 2024
```

## Including one additional `.qmd` file with **only** three `flextable` objects and no `ggplot` objects
```
$ time quarto1.3 render
...
real    1m43,202s
user    0m0,000s
sys     0m0,031s
```

```
$ time quarto render
...
real    6m54,161s
user    0m0,062s
sys     0m0,015s
```
Yet again, it hangs before the post-render script starts.

---
Any ideas as to what could cause this difference? The problem seems to be somehow linked to `flextable` and I recall that rendering a PDF book with `flextables` worked again from Quarto 1.4.551 onwards thanks to #8843, but I can't really grasp how the interaction changes despite having the same version of `flextable`. I am using `flextable==0.9.5`.

Thank you in advance!

**Comment by mcanouil** [comment]
Closing this as outdated!

- #9729
